Invalid child routes in AlertmanagerConfig CRD triggers a panic

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

## Description

The prometheus operator crashes on a bad CRD

## Steps to Reproduce

Using this:

```yaml
---
apiVersion: monitoring.coreos.com/v1alpha1
kind: AlertmanagerConfig
metadata:
  name: global-config
  labels:
    role: alert-rules
    namespace: monitoring
spec:
  route:
    groupBy:
      - namespace
      - alertname
    groupWait: 30s
    groupInterval: 5m
    repeatInterval: 12h
    receiver: 'matrix'
    # All alerts that do not match the following child routes
    # will remain at the root node and be dispatched to 'matrix'.
    routes:
      - receiver: 'null'
        matchers: [ alertname = "Watchdog" ]
```

## Expected Result

Ideally, the api server would reject the above CRD when trying to apply it to the cluster

## Actual Result

The CRD was expected, because the specification of the CRD is lax:

https://github.com/prometheus-operator/prometheus-operator/blob/372b00558d06a0aa2bc5f544f5c754dff4872527/example/prometheus-operator-crd/monitoring.coreos.com_alertmanagerconfigs.yaml#L5710-L5714




### Prometheus Operator Version

```shell
Name:                   kube-prometheus-stack-operator                                                                                                                                                             
Namespace:              monitoring                                                                                                                                                                                 
CreationTimestamp:      Fri, 20 Jan 2023 14:29:10 -0500                                                                                                                                                            
Labels:                 app=kube-prometheus-stack-operator                                                                                                                                                         
                        app.kubernetes.io/component=prometheus-operator                                                                                                                                            
                        app.kubernetes.io/instance=kube-prometheus-stack                                                                                                                                           
                        app.kubernetes.io/managed-by=Helm                                                                                                                                                          
                        app.kubernetes.io/name=kube-prometheus-stack-prometheus-operator                                                                                                                           
                        app.kubernetes.io/part-of=kube-prometheus-stack                                                                                                                                            
                        app.kubernetes.io/version=56.2.1                                                                                                                                                           
                        chart=kube-prometheus-stack-56.2.1                                                                                                                                                         
                        heritage=Helm                                                                                                                                                                              
                        release=kube-prometheus-stack                                                                                                                                                              
Annotations:            deployment.kubernetes.io/revision: 41
```


### Kubernetes Version

```yaml
kubectl version -o yaml
clientVersion:
  buildDate: "2022-06-15T14:22:29Z"
  compiler: gc
  gitCommit: f66044f4361b9f1f96f0053dd46cb7dce5e990a8
  gitTreeState: clean
  gitVersion: v1.24.2
  goVersion: go1.18.3
  major: "1"
  minor: "24"
  platform: linux/amd64
kustomizeVersion: v4.5.4
serverVersion:
  buildDate: "2023-12-27T16:01:02Z"
  compiler: gc
  gitCommit: 5b2d1271a6a00a8d71981cc968bc0f822620b9d8
  gitTreeState: clean
  gitVersion: v1.28.5+k3s1
  goVersion: go1.20.12
  major: "1"
  minor: "28"
  platform: linux/amd64
```


### Kubernetes Cluster Type

Rancher

### How did you deploy Prometheus-Operator?

helm chart:prometheus-community/kube-prometheus-stack

Version: 0.71.2

### Manifests

_No response_

### prometheus-operator log output

```shell
panic: route[0]: json: cannot unmarshal string into Go struct field Route.matchers of type v1alpha1.Matcher
```


### Anything else?

_No response_

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,5 +1,6 @@
 ## 0.76.1 / 2025-09-03
 
+* [BUGFIX] fix panic when processing an invalid `AlertmanagerConfig` object used for global configuration. #6931
 * [BUGFIX] fix bug with Kubernetes service discovery `Selector.Role` field. #6896
 
 ## 0.76.0 / 2025-08-08

--- a/pkg/alertmanager/amcfg.go
+++ b/pkg/alertmanager/amcfg.go
@@ -247,6 +247,10 @@ func (cb *configBuilder) initializeFromAlertmanagerConfig(ctx context.Context, g
 		Name:      amConfig.Name,
 	}
 
+	if err := checkAlertmanagerConfigResource(ctx, amConfig, cb.amVersion, cb.store); err != nil {
+		return err
+	}
+
 	global, err := cb.convertGlobalConfig(ctx, globalConfig, crKey)
 	if err != nil {
 		return err

--- a/pkg/alertmanager/amcfg_test.go
+++ b/pkg/alertmanager/amcfg_test.go
@@ -145,6 +145,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 						{
 							Name: "null",
 						},
+						{
+							Name: "myreceiver",
+						},
 					},
 					Route: &monitoringv1alpha1.Route{
 						Receiver: "null",
@@ -190,6 +193,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 					{
 						Name: "mynamespace/global-config/null",
 					},
+					{
+						Name: "mynamespace/global-config/myreceiver",
+					},
 				},
 				Route: &route{
 					Receiver: "mynamespace/global-config/null",
@@ -225,6 +231,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 						{
 							Name: "null",
 						},
+						{
+							Name: "myreceiver",
+						},
 					},
 					Route: &monitoringv1alpha1.Route{
 						Receiver: "null",
@@ -246,6 +255,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 				Receivers: []*receiver{
 					{
 						Name: "mynamespace/global-config/null",
+					},
+					{
+						Name: "mynamespace/global-config/myreceiver",
 					},
 				},
 				Route: &route{
@@ -354,6 +366,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 						{
 							Name: "null",
 						},
+						{
+							Name: "myreceiver",
+						},
 					},
 					Route: &monitoringv1alpha1.Route{
 						Receiver: "null",
@@ -375,6 +390,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 				Receivers: []*receiver{
 					{
 						Name: "mynamespace/global-config/null",
+					},
+					{
+						Name: "mynamespace/global-config/myreceiver",
 					},
 				},
 				Route: &route{
@@ -483,6 +501,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 						{
 							Name: "null",
 						},
+						{
+							Name: "myreceiver",
+						},
 					},
 					Route: &monitoringv1alpha1.Route{
 						Receiver: "null",
@@ -504,6 +525,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 				Receivers: []*receiver{
 					{
 						Name: "mynamespace/global-config/null",
+					},
+					{
+						Name: "mynamespace/global-config/myreceiver",
 					},
 				},
 				Route: &route{
@@ -571,6 +595,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 						{
 							Name: "null",
 						},
+						{
+							Name: "myreceiver",
+						},
 					},
 					Route: &monitoringv1alpha1.Route{
 						Receiver: "null",
@@ -592,6 +619,9 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 				Receivers: []*receiver{
 					{
 						Name: "mynamespace/global-config/null",
+					},
+					{
+						Name: "mynamespace/global-config/myreceiver",
 					},
 				},
 				Route: &route{
@@ -863,6 +893,34 @@ func TestInitializeFromAlertmanagerConfig(t *testing.T) {
 				},
 			},
 			wantErr: false,
+		},
+		{
+			name: "invalid alertmanagerConfig with invalid child routes",
+			amConfig: &monitoringv1alpha1.AlertmanagerConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "global-config",
+					Namespace: "mynamespace",
+				},
+				Spec: monitoringv1alpha1.AlertmanagerConfigSpec{
+					Receivers: []monitoringv1alpha1.Receiver{
+						{
+							Name: "null",
+						},
+					},
+					Route: &monitoringv1alpha1.Route{
+						Receiver: "null",
+						Routes: []apiextensionsv1.JSON{
+							{
+								Raw: []byte(`{"receiver": "recv2", "matchers": [{"severity":"!=critical$"}]}`),
+							},
+						},
+					},
+				},
+			},
+			matcherStrategy: monitoringingv1.AlertmanagerConfigMatcherStrategy{
+				Type: "OnNamespace",
+			},
+			wantErr: true,
 		},
 	}
 	for _, tt := range tests {

--- a/pkg/alertmanager/operator.go
+++ b/pkg/alertmanager/operator.go
@@ -1062,10 +1062,14 @@ func (c *Operator) selectAlertmanagerConfigs(ctx context.Context, am *monitoring
 // checkAlertmanagerConfigResource verifies that an AlertmanagerConfig object is valid
 // for the given Alertmanager version and has no missing references to other objects.
 func checkAlertmanagerConfigResource(ctx context.Context, amc *monitoringv1alpha1.AlertmanagerConfig, amVersion semver.Version, store *assets.StoreBuilder) error {
+	// Perform semantic validation irrespective of the Alertmanager version.
 	if err := validationv1alpha1.ValidateAlertmanagerConfig(amc); err != nil {
 		return err
 	}
 
+	// Perform more specific validations which depend on the Alertmanager
+	// version. It also retrieves data from referenced secrets and configmaps
+	// (and fails in case of missing/invalid references).
 	if err := checkReceivers(ctx, amc, store, amVersion); err != nil {
 		return err
 	}

--- a/pkg/alertmanager/operator_test.go
+++ b/pkg/alertmanager/operator_test.go
@@ -191,9 +191,9 @@ func TestCreateStatefulSetInputHash(t *testing.T) {
 
 // Test to exercise the function checkAlertmanagerConfigResource
 // and validate that semantic validation is in place for all the fields in the
-// AlertmanagerConfig CR. The validation is preformed by the operator
-// after selecting AlertmanagerConfig resources but before passing them to
-// addAlertmanagerConfigs.
+// AlertmanagerConfig CR. The validation is performed by the operator
+// after selecting AlertmanagerConfig resources and before generating the
+// Alertmanager configuration.
 func TestCheckAlertmanagerConfig(t *testing.T) {
 	version, err := semver.ParseTolerant(operator.DefaultAlertmanagerVersion)
 	require.NoError(t, err)
@@ -958,15 +958,131 @@ func TestCheckAlertmanagerConfig(t *testing.T) {
 			},
 			ok: true,
 		},
+		{
+			amConfig: &monitoringv1alpha1.AlertmanagerConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "subroute-with-unknow-field",
+					Namespace: "ns1",
+				},
+				Spec: monitoringv1alpha1.AlertmanagerConfigSpec{
+					Route: &monitoringv1alpha1.Route{
+						Receiver: "recv1",
+						Routes: []apiextensionsv1.JSON{
+							{
+								Raw: []byte(`{"receiver": "recv2", "matchers": [{"severity":"!=critical$"}]}`),
+							},
+						},
+					},
+					Receivers: []monitoringv1alpha1.Receiver{{
+						Name: "recv1",
+					}, {
+						Name: "recv2",
+					}},
+				},
+			},
+		},
+		{
+			amConfig: &monitoringv1alpha1.AlertmanagerConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "subroute-with-invalid-matcher",
+					Namespace: "ns1",
+				},
+				Spec: monitoringv1alpha1.AlertmanagerConfigSpec{
+					Route: &monitoringv1alpha1.Route{
+						Receiver: "recv1",
+						Routes: []apiextensionsv1.JSON{
+							{
+								Raw: []byte(`{"receiver": "recv2", "matchers": [{"name": "severity", "value": "critical", "matchType": "!!"}]}`),
+							},
+						},
+					},
+					Receivers: []monitoringv1alpha1.Receiver{{
+						Name: "recv1",
+					}, {
+						Name: "recv2",
+					}},
+				},
+			},
+		},
+		{
+			amConfig: &monitoringv1alpha1.AlertmanagerConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "subroute-with-empty-matcher-name",
+					Namespace: "ns1",
+				},
+				Spec: monitoringv1alpha1.AlertmanagerConfigSpec{
+					Route: &monitoringv1alpha1.Route{
+						Receiver: "recv1",
+						Routes: []apiextensionsv1.JSON{
+							{
+								Raw: []byte(`{"receiver": "recv2", "matchers": [{"name": "", "value": "critical", "matchType": "!="}]}`),
+							},
+						},
+					},
+					Receivers: []monitoringv1alpha1.Receiver{{
+						Name: "recv1",
+					}, {
+						Name: "recv2",
+					}},
+				},
+			},
+		},
+		{
+			amConfig: &monitoringv1alpha1.AlertmanagerConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "subroute-with-missing-receiver",
+					Namespace: "ns1",
+				},
+				Spec: monitoringv1alpha1.AlertmanagerConfigSpec{
+					Route: &monitoringv1alpha1.Route{
+						Receiver: "recv1",
+						Routes: []apiextensionsv1.JSON{
+							{
+								Raw: []byte(`{"receiver": "recv2", "matchers": [{"name": "severity", "value": "critical", "matchType": "!="}]}`),
+							},
+						},
+					},
+					Receivers: []monitoringv1alpha1.Receiver{{
+						Name: "recv1",
+					}},
+				},
+			},
+		},
+		{
+			amConfig: &monitoringv1alpha1.AlertmanagerConfig{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "valid-subroute-definition",
+					Namespace: "ns1",
+				},
+				Spec: monitoringv1alpha1.AlertmanagerConfigSpec{
+					Route: &monitoringv1alpha1.Route{
+						Receiver: "recv1",
+						Routes: []apiextensionsv1.JSON{
+							{
+								Raw: []byte(`{"receiver": "recv2", "matchers": [{"name": "severity", "value": "critical", "matchType": "!="}]}`),
+							},
+						},
+					},
+					Receivers: []monitoringv1alpha1.Receiver{{
+						Name: "recv1",
+					}, {
+						Name: "recv2",
+					}},
+				},
+			},
+			ok: true,
+		},
 	} {
 		t.Run(tc.amConfig.Name, func(t *testing.T) {
 			store := assets.NewStoreBuilder(c.CoreV1(), c.CoreV1())
+
 			err := checkAlertmanagerConfigResource(context.Background(), tc.amConfig, version, store)
 			if tc.ok {
 				require.NoError(t, err)
 				return
 			}
 
+			t.Logf("err: %s", err)
 			require.Error(t, err)
 		})
 	}

--- a/pkg/alertmanager/validation/v1alpha1/validation.go
+++ b/pkg/alertmanager/validation/v1alpha1/validation.go
@@ -42,7 +42,7 @@ func ValidateAlertmanagerConfig(amc *monitoringv1alpha1.AlertmanagerConfig) erro
 		return err
 	}
 
-	return validateAlertManagerRoutes(amc.Spec.Route, receivers, muteTimeIntervals, true)
+	return validateRoute(amc.Spec.Route, receivers, muteTimeIntervals, true)
 }
 
 func validateReceivers(receivers []monitoringv1alpha1.Receiver) (map[string]struct{}, error) {
@@ -345,10 +345,11 @@ func validateMSTeamsConfigs(configs []monitoringv1alpha1.MSTeamsConfig) error {
 	return nil
 }
 
-// validateAlertManagerRoutes verifies that the given route and all its children are semantically valid.
-// because of the self-referential issues mentioned in https://github.com/kubernetes/kubernetes/issues/62872
-// it is not currently possible to apply OpenAPI validation to a v1alpha1.Route.
-func validateAlertManagerRoutes(r *monitoringv1alpha1.Route, receivers, muteTimeIntervals map[string]struct{}, topLevelRoute bool) error {
+// validateRoute verifies that the given route and all its children are
+// semantically valid.  because of the self-referential issues mentioned in
+// https://github.com/kubernetes/kubernetes/issues/62872 it is not currently
+// possible to apply OpenAPI validation to a v1alpha1.Route.
+func validateRoute(r *monitoringv1alpha1.Route, receivers, muteTimeIntervals map[string]struct{}, topLevelRoute bool) error {
 	if r == nil {
 		return nil
 	}
@@ -388,7 +389,6 @@ func validateAlertManagerRoutes(r *monitoringv1alpha1.Route, receivers, muteTime
 		}
 	}
 
-	// validate that if defaults are set, they match regex
 	if r.GroupInterval != "" && !durationRe.MatchString(r.GroupInterval) {
 		return fmt.Errorf("groupInterval %s does not match required regex: %s", r.GroupInterval, durationRe.String())
 
@@ -401,13 +401,20 @@ func validateAlertManagerRoutes(r *monitoringv1alpha1.Route, receivers, muteTime
 		return fmt.Errorf("repeatInterval %s does not match required regex: %s", r.RepeatInterval, durationRe.String())
 	}
 
+	for i, m := range r.Matchers {
+		if err := m.Validate(); err != nil {
+			return fmt.Errorf("matcher[%d]: %w", i, err)
+		}
+	}
+
+	// Unmarshal the child routes and validate them recursively.
 	children, err := r.ChildRoutes()
 	if err != nil {
 		return err
 	}
 
 	for i := range children {
-		if err := validateAlertManagerRoutes(&children[i], receivers, muteTimeIntervals, false); err != nil {
+		if err := validateRoute(&children[i], receivers, muteTimeIntervals, false); err != nil {
 			return fmt.Errorf("route[%d]: %w", i, err)
 		}
 	}

--- a/pkg/alertmanager/validation/v1beta1/validation.go
+++ b/pkg/alertmanager/validation/v1beta1/validation.go
@@ -42,7 +42,7 @@ func ValidateAlertmanagerConfig(amc *monitoringv1beta1.AlertmanagerConfig) error
 		return err
 	}
 
-	return validateAlertManagerRoutes(amc.Spec.Route, receivers, timeIntervals, true)
+	return validateRoute(amc.Spec.Route, receivers, timeIntervals, true)
 }
 
 func validateReceivers(receivers []monitoringv1beta1.Receiver) (map[string]struct{}, error) {
@@ -346,10 +346,11 @@ func validateMSTeamsConfigs(configs []monitoringv1beta1.MSTeamsConfig) error {
 	return nil
 }
 
-// validateAlertManagerRoutes verifies that the given route and all its children are semantically valid.
-// because of the self-referential issues mentioned in https://github.com/kubernetes/kubernetes/issues/62872
-// it is not currently possible to apply OpenAPI validation to a v1beta1.Route.
-func validateAlertManagerRoutes(r *monitoringv1beta1.Route, receivers, timeIntervals map[string]struct{}, topLevelRoute bool) error {
+// validateRoute verifies that the given route and all its children are
+// semantically valid.  because of the self-referential issues mentioned in
+// https://github.com/kubernetes/kubernetes/issues/62872 it is not currently
+// possible to apply OpenAPI validation to a v1beta1.Route.
+func validateRoute(r *monitoringv1beta1.Route, receivers, timeIntervals map[string]struct{}, topLevelRoute bool) error {
 	if r == nil {
 		return nil
 	}
@@ -389,7 +390,6 @@ func validateAlertManagerRoutes(r *monitoringv1beta1.Route, receivers, timeInter
 		}
 	}
 
-	// validate that if defaults are set, they match regex
 	if r.GroupInterval != "" && !durationRe.MatchString(r.GroupInterval) {
 		return fmt.Errorf("groupInterval %s does not match required regex: %s", r.GroupInterval, durationRe.String())
 
@@ -402,13 +402,20 @@ func validateAlertManagerRoutes(r *monitoringv1beta1.Route, receivers, timeInter
 		return fmt.Errorf("repeatInterval %s does not match required regex: %s", r.RepeatInterval, durationRe.String())
 	}
 
+	for i, v := range r.Matchers {
+		if err := v.Validate(); err != nil {
+			return fmt.Errorf("matcher[%d]: %w", i, err)
+		}
+	}
+
+	// Unmarshal the child routes and validate them recursively.
 	children, err := r.ChildRoutes()
 	if err != nil {
 		return err
 	}
 
 	for i := range children {
-		if err := validateAlertManagerRoutes(&children[i], receivers, timeIntervals, false); err != nil {
+		if err := validateRoute(&children[i], receivers, timeIntervals, false); err != nil {
 			return fmt.Errorf("route[%d]: %w", i, err)
 		}
 	}

--- a/pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
+++ b/pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go
@@ -15,6 +15,7 @@
 package v1alpha1
 
 import (
+	"bytes"
 	"encoding/json"
 	"errors"
 	"fmt"
@@ -146,7 +147,9 @@ func (r *Route) ChildRoutes() ([]Route, error) {
 	out := make([]Route, len(r.Routes))
 
 	for i, v := range r.Routes {
-		if err := json.Unmarshal(v.Raw, &out[i]); err != nil {
+		dec := json.NewDecoder(bytes.NewBuffer(v.Raw))
+		dec.DisallowUnknownFields()
+		if err := dec.Decode(&out[i]); err != nil {
 			return nil, fmt.Errorf("route[%d]: %w", i, err)
 		}
 	}

--- a/pkg/apis/monitoring/v1beta1/alertmanager_config_types.go
+++ b/pkg/apis/monitoring/v1beta1/alertmanager_config_types.go
@@ -15,6 +15,7 @@
 package v1beta1
 
 import (
+	"bytes"
 	"encoding/json"
 	"errors"
 	"fmt"
@@ -143,7 +144,9 @@ func (r *Route) ChildRoutes() ([]Route, error) {
 	out := make([]Route, len(r.Routes))
 
 	for i, v := range r.Routes {
-		if err := json.Unmarshal(v.Raw, &out[i]); err != nil {
+		dec := json.NewDecoder(bytes.NewBuffer(v.Raw))
+		dec.DisallowUnknownFields()
+		if err := dec.Decode(&out[i]); err != nil {
 			return nil, fmt.Errorf("route[%d]: %w", i, err)
 		}
 	}

